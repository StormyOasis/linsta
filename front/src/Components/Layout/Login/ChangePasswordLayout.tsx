import React, { useState } from "react";
import { styled } from "styled-components";
import { Navigate, useSearchParams } from 'react-router-dom';
import * as styles from "../../Common/CombinedStyling";
import Theme from "../../../Components/Themes/Theme";
import StyledInput from "../../../Components/Common/StyledInput";
import { validatePassword } from "../../../utils/utils";
import StyledButton from "../../../Components/Common/StyledButton";
import { postChangePassword } from "../../../api/ServiceController";
import { getCurrentUser } from "../../../api/Auth";

type ChangeProps = {

};

const ChangeLayoutWrapper = styled.main<any>`
  display: flex;
  flex-direction: column;
  flex-grow: 1;
`;

const ChangeHeaderImage = styled.span`
    background-image: url('/public/images/forgot.svg');
    display: block; 
    height: 133px; 
    width: 133px;
`;

const HeadingText = styled.div`
    margin-top: 10px;
    text-align:center;
    font-weight: 600;        
`;

const SubHeadingText = styled.div`
    color: rgb(115,115,155);
    text-align: center;
    margin-top:6px;
    margin-bottom: 10px;
`;

const ChangeForm = styled.form`
  display: flex;
  flex-direction: column;
  margin: 0;
  max-width: 350px;
  vertical-align: baseline;
`;

const isFormValid = (password1: string, password2: string, token?: string|null, currentPassword?: string) => {
    if(password1 != password2) {
        return false;
    }
    
    if(!validatePassword(password1)) {
        return false;
    }

    if(token == null && currentPassword?.length === 0) {        
        return false;
    }

    return true;
}

const onSubmit = async (setIsFinished:any, password1: string, password2: string, token?: string|null, currentPassword?: string) => {
    try {
        const result = await postChangePassword({userName: getCurrentUser()?.userName, oldPassword: currentPassword, token, password1, password2});
        
        if(result.status == 200 && result.data.status === "OK") {
            setIsFinished(true);
        }        
        else {
            throw new Error();
        }        

    } catch(err) {
        console.error(err);
        return false;
    }
    return true;
}

const renderChangeForm = () => {
    const [currentPassword, setCurrentPassword] = useState("");
    const [password1, setPassword1] = useState("");
    const [password2, setPassword2] = useState("");
    const [isFinished, setIsFinished] = useState(false);

    const [searchParams, _setSearchParams] = useSearchParams();

    const token = searchParams.get('token');
    return (
        <>
            {isFinished && <Navigate to="/" replace={true} />}
            <ChangeHeaderImage/>
            <ChangeForm method="post">
                <HeadingText>Reset Password</HeadingText>
                <SubHeadingText>
                    Your password must be between 8 and 15 characters and should include both upper and lower case letters, numbers, and special characters.
                </SubHeadingText>
                {token == null &&                  
                    <StyledInput datatestid='currentPassword'
                        name="currentPassword"
                        placeholder="Current Password"
                        value={currentPassword}
                        type="password"
                        isValid={validatePassword(currentPassword)}
                        onChange={(event: React.ChangeEvent<HTMLInputElement>) => setCurrentPassword(event.target.value)} />
                }
                <StyledInput
                    name="password1"
                    placeholder="New Password"
                    value={password1}
                    type="password"
                    isValid={validatePassword(password1)}
                    onChange={(event: React.ChangeEvent<HTMLInputElement>) => setPassword1(event.target.value)} />       
                <StyledInput
                    name="password2"
                    placeholder="Retype Password"
                    type="password"
                    value={password2}
                    isValid={validatePassword(password2)}
                    onChange={(event: React.ChangeEvent<HTMLInputElement>) => setPassword2(event.target.value)} />                         
                <StyledButton
                    type="button"
                    disabled={!isFormValid(password1, password2, token, currentPassword)}
                    onClick={() => {return onSubmit(setIsFinished, password1, password2, token, currentPassword);}}
                    text="Change Password">
                </StyledButton>

            </ChangeForm>        
        </>
    );
}

const ChangePasswordLayout: React.FC<ChangeProps> = (_props: ChangeProps) => {
    return (
        <Theme>
            <ChangeLayoutWrapper role="main">
                <div className={styles.default.innerDiv1}>
                    <div className={styles.default.innerDiv2}>
                        <div className={styles.default.signupBox}>
                            {renderChangeForm()}
                        </div>
                    </div>
                </div>
            </ChangeLayoutWrapper>
        </Theme>
    );
}

export default ChangePasswordLayout;