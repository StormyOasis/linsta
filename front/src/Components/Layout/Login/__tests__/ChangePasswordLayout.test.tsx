import React from 'react'
import {render, screen} from '@testing-library/react';
import '@testing-library/jest-dom';
import ChangePasswordLayout from '../ChangePasswordLayout';
import { MemoryRouter } from 'react-router-dom';

describe("ChangePasswordLayout Component", () => {
    it("should render using current password as input", () => {
        const r = render(<MemoryRouter><ChangePasswordLayout></ChangePasswordLayout></MemoryRouter>).asFragment();
        expect(r).toMatchSnapshot();
        expect(screen.getByTestId("currentPassword")).toBeInTheDocument();
    });

    it("should render based on token", () => {
        const r = render(<MemoryRouter initialEntries={[`?token="1234"`]}><ChangePasswordLayout></ChangePasswordLayout></MemoryRouter>).asFragment();

        expect(r).toMatchSnapshot();
        expect(screen.queryByTestId("currentPassword")).toBeNull();
    });
});