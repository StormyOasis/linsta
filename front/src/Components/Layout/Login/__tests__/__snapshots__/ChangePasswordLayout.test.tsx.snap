// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChangePasswordLayout Component should render based on token 1`] = `
<DocumentFragment>
  <main
    class="ChangePasswordLayout__ChangeLayoutWrapper-sc-1lk64wu-0 dfcMAT"
    role="main"
  >
    <div>
      <div>
        <div>
          <span
            class="ChangePasswordLayout__ChangeHeaderImage-sc-1lk64wu-1 gbVSxh"
          />
          <form
            class="ChangePasswordLayout__ChangeForm-sc-1lk64wu-4 gPcaiU"
            method="post"
          >
            <div
              class="ChangePasswordLayout__HeadingText-sc-1lk64wu-2 jIZyGa"
            >
              Reset Password
            </div>
            <div
              class="ChangePasswordLayout__SubHeadingText-sc-1lk64wu-3 kiigiu"
            >
              Your password must be between 8 and 15 characters and should include both upper and lower case letters, numbers, and special characters.
            </div>
            <div
              class="StyledInput__StyledInputWrapper-sc-fkf055-0 hOmQjA"
            >
              <label
                class="StyledInput__StyledInputLabel-sc-fkf055-2 gizmBW"
              >
                <input
                  class="StyledInput__StyledInputInput-sc-fkf055-1 ejdHNL"
                  maxlength="255"
                  name="password1"
                  placeholder="New Password"
                  type="password"
                  value=""
                />
              </label>
            </div>
            <div
              class="StyledInput__StyledInputWrapper-sc-fkf055-0 hOmQjA"
            >
              <label
                class="StyledInput__StyledInputLabel-sc-fkf055-2 gizmBW"
              >
                <input
                  class="StyledInput__StyledInputInput-sc-fkf055-1 ejdHNL"
                  maxlength="255"
                  name="password2"
                  placeholder="Retype Password"
                  type="password"
                  value=""
                />
              </label>
            </div>
            <button
              class="StyledButton__StyledButtonComponent-sc-8t4kac-0 fLkIsR"
              disabled=""
              type="button"
            >
              Change Password
            </button>
          </form>
        </div>
      </div>
    </div>
  </main>
</DocumentFragment>
`;

exports[`ChangePasswordLayout Component should render using current password as input 1`] = `
<DocumentFragment>
  <main
    class="ChangePasswordLayout__ChangeLayoutWrapper-sc-1lk64wu-0 dfcMAT"
    role="main"
  >
    <div>
      <div>
        <div>
          <span
            class="ChangePasswordLayout__ChangeHeaderImage-sc-1lk64wu-1 gbVSxh"
          />
          <form
            class="ChangePasswordLayout__ChangeForm-sc-1lk64wu-4 gPcaiU"
            method="post"
          >
            <div
              class="ChangePasswordLayout__HeadingText-sc-1lk64wu-2 jIZyGa"
            >
              Reset Password
            </div>
            <div
              class="ChangePasswordLayout__SubHeadingText-sc-1lk64wu-3 kiigiu"
            >
              Your password must be between 8 and 15 characters and should include both upper and lower case letters, numbers, and special characters.
            </div>
            <div
              class="StyledInput__StyledInputWrapper-sc-fkf055-0 hOmQjA"
            >
              <label
                class="StyledInput__StyledInputLabel-sc-fkf055-2 gizmBW"
              >
                <input
                  class="StyledInput__StyledInputInput-sc-fkf055-1 ejdHNL"
                  data-testid="currentPassword"
                  maxlength="255"
                  name="currentPassword"
                  placeholder="Current Password"
                  type="password"
                  value=""
                />
              </label>
            </div>
            <div
              class="StyledInput__StyledInputWrapper-sc-fkf055-0 hOmQjA"
            >
              <label
                class="StyledInput__StyledInputLabel-sc-fkf055-2 gizmBW"
              >
                <input
                  class="StyledInput__StyledInputInput-sc-fkf055-1 ejdHNL"
                  maxlength="255"
                  name="password1"
                  placeholder="New Password"
                  type="password"
                  value=""
                />
              </label>
            </div>
            <div
              class="StyledInput__StyledInputWrapper-sc-fkf055-0 hOmQjA"
            >
              <label
                class="StyledInput__StyledInputLabel-sc-fkf055-2 gizmBW"
              >
                <input
                  class="StyledInput__StyledInputInput-sc-fkf055-1 ejdHNL"
                  maxlength="255"
                  name="password2"
                  placeholder="Retype Password"
                  type="password"
                  value=""
                />
              </label>
            </div>
            <button
              class="StyledButton__StyledButtonComponent-sc-8t4kac-0 fLkIsR"
              disabled=""
              type="button"
            >
              Change Password
            </button>
          </form>
        </div>
      </div>
    </div>
  </main>
</DocumentFragment>
`;
